drivers/java/src/main/java/com/rethinkdb/net/SocketWrapper.java

```diff
@@ -2,47 +2,86 @@
+import java.io.OutputStream;
 import java.net.InetSocketAddress;
+import java.net.Socket;
 import java.net.SocketException;
 import java.net.SocketTimeoutException;
 import java.nio.ByteBuffer;
-import java.nio.channels.SocketChannel;
-import java.net.StandardSocketOptions;
+import java.nio.ByteOrder;
 import java.nio.charset.StandardCharsets;
-import java.util.*;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Optional;
 import java.util.concurrent.TimeoutException;
+import java.util.concurrent.locks.ReentrantLock;
 
 public class SocketWrapper {
-    private SocketChannel socketChannel;
-    private Optional<Long> timeout = Optional.empty();
+    final ReentrantLock mu = new ReentrantLock();
+
+    private Socket socket = null;
+
+    private SocketFactory socketFactory = SocketFactory.getDefault();
+    private SSLSocket sslSocket = null;
 
-    public final String hostname;
-    public final int port;
+    protected OutputStream writeStream = null;
+    protected InputStream readStream = null;
 
-    public SocketWrapper(String hostname, int port,
+    // options
+    private Optional<SSLContext> sslContext = Optional.empty();
+    private Optional<Long> timeout = Optional.empty();
+    private final String hostname;
+    private final int port;
+
+    public SocketWrapper(String hostname,
+                         int port,
+                         Optional<SSLContext> sslContext,
                          Optional<Long> timeout) {
         this.hostname = hostname;
         this.port = port;
+        this.sslContext = sslContext;
         this.timeout = timeout;
-        try {
-            this.socketChannel = SocketChannel.open();
-        } catch (IOException e) {
-            throw new ReqlDriverError(e);
-        }
     }
 
     public void connect(ByteBuffer handshake) throws TimeoutException {
         Optional<Long> deadline = timeout.map(Util::deadline);
+
+        mu.lock();
         try {
-            socketChannel.configureBlocking(true);
-            socketChannel.setOption(StandardSocketOptions.TCP_NODELAY, true);
-            socketChannel.setOption(StandardSocketOptions.SO_KEEPALIVE, true);
-            socketChannel.socket().connect(
-                    new InetSocketAddress(hostname, port),
-                    timeout.orElse(0L).intValue());
-            socketChannel.write(handshake);
-            String msg = readNullTerminatedString(deadline);
+            // establish connection
+            final InetSocketAddress addr = new InetSocketAddress(hostname, port);
+            socket = socketFactory.createSocket();
+            socket.connect(addr, timeout.orElse(0L).intValue());
+            socket.setTcpNoDelay(true);
+            socket.setKeepAlive(true);
+
+            // should we secure the connection?
+            if (sslContext.isPresent()){
+                socketFactory = sslContext.get().getSocketFactory();
+                SSLSocketFactory sslSf = (SSLSocketFactory) socketFactory;
```

Comment:
`socketFactory` is declared as type `SocketFactory` and used in the code for non-SSL sockets.